LLM Translation Integration with Ollama

hyattc1 opened the pull request:
## Summary
This PR integrates LLM-powered translation directly into NodeBB's backend, enabling automatic detection and translation of non-English posts using a local Ollama instance with the Mistral model.

## Changes Made

### New Files
- **`src/llm/translate.js`** - Core translation module with:
  - Multi-strategy language detection (non-ASCII characters + common word matching)
  - Direct Ollama API integration via Node.js `fetch`
  - Robust error handling with fallback to original text
  - Support for Asian languages (Chinese, Japanese, Arabic) and European languages (Spanish, French, German)

### Modified Files
- **`src/translate/index.js`** - Updated to use new LLM module instead of stub implementation

## Features

### Language Detection
Uses a two-pronged approach:
1. **Non-ASCII character analysis** - Catches Asian languages (>10% non-ASCII characters)
2. **Common word matching** - Detects Spanish, French, German words (e.g., "hola", "como", "bonjour", "merci")

### Translation Flow
1. Post creation triggers `translate.translate()`
2. Language detection determines if text is English
3. Non-English text is sent to Ollama with prompt: `"Translate the following text into English:\n<text>"`
4. Translation stored in `translatedContent` field
5. UI displays translate button on non-English posts
6. Click to toggle visibility of translation

### Error Handling
- Connection failures fall back to marking post as English
- Translation errors return `[Translation unavailable] <original text>`
- All errors logged to console for debugging

## Configuration

### Required Setup
Teammates need to run Ollama locally:
ershell
# PowerShell (Windows)
$env:OLLAMA_HOST="0.0.0.0:11434"
ollama serve
# Bash (Linux/Mac)
export OLLAMA_HOST="0.0.0.0:11434"
ollama serve### Config File
The `config.json` should include (not committed, local only):
{
  "translateApiUrl": "http://host.docker.internal:11434/api/generate"
}**Note**: `host.docker.internal` works automatically in Codespaces/Docker environments to reach the host machine's Ollama instance.

## Testing

### Test Cases
1. **English post**: `"Hello, this is a test"` → No translate button
2. **Spanish post**: `"Hola, ¿cómo estás?"` → Translate button appears
3. **Chinese post**: `"你好，这是一个测试"` → Translate button appears
4. **Japanese post**: `"今日はとても疲れています"` → Translate button appears

### Manual Testing Steps
1. Start Ollama: `ollama serve` with `OLLAMA_HOST=0.0.0.0:11434`
2. Start NodeBB: `./nodebb start`
3. Create a post with non-English text
4. Verify translate button appears
5. Click button to view translation

## Architecture

### Backend-Only Implementation
- No UI changes required (uses existing translate button)
- No new routes or API endpoints
- Translation happens at post creation time
- Integrates seamlessly with existing `src/posts/create.js` flow

### Data Flow

hyattc1 committed:
Add LLM translation integration with Ollama

- Created src/llm/translate.js with language detection and Ollama API integration
- Updated src/translate/index.js to use new LLM module
- Language detection uses multi-strategy approach:
  * Non-ASCII characters for Asian languages
  * Common word matching for European languages (Spanish, French, German)
- Direct Ollama integration via fetch API
- Robust error handling with fallback to original text
- Works with existing UI (translate button on posts)
- Config: translateApiUrl = http://host.docker.internal:11434/api/generate

hyattc1 committed:
Fix linting errors in translate.js

- Replace ++ with += 1 (no-plusplus)
- Add trailing commas to arrays (@stylistic/js/comma-dangle)

charliez07 reviewed (commented):
The code changes look good oevrall. The implementation addresses the required features.

zsrobinson84 commented:
Code looks good! I like how you implemented and adjusted the code from our collab file (as we discussed in person). I downloaded the file, but when I hit translate,  I am getting Object object. I don't know what is causing this based on our code - might have to take another look at our last meeting

davecrunch reviewed (commented):
Sometimes the llm translation output produces more than just the translation. Maybe we can change the prompt on line 99 of translate.js to be more specific? Like explicitly tell the model to only give the translation and nothing else. Otherwise, the logic looks good!

hyattc1 commented:
> Code looks good! I like how you implemented and adjusted the code from our collab file (as we discussed in person). I downloaded the file, but when I hit translate, I am getting Object object. I don't know what is causing this based on our code - might have to take another look at our last meeting

It’s not the code. It works on my side because Ollama is running. On your machine the microservice can’t reach Ollama, so it sends back a default object, and the browser just prints it as [object Object]. Start Ollama and make sure the model is loaded, then hit Translate again. It should show the real text once Ollama is up.

hyattc1 commented:
> Sometimes the llm translation output produces more than just the translation. Maybe we can change the prompt on line 99 of translate.js to be more specific? Like explicitly tell the model to only give the translation and nothing else. Otherwise, the logic looks good!

Good point. We could tighten the prompt so it tells the model to output only the translation and nothing else. That should stop the extra lines. Everything else in the logic looks solid.

hyattc1 committed:
Improve translation prompt to output only translated text

Updated prompt to explicitly instruct the model to output ONLY the translated text with no explanations, formatting, or commentary. This prevents the model from adding extra information like translation notes or context.
